packages/@aws-cdk/core/lib/cfn-parse.ts

```diff
@@ -67,8 +67,16 @@ export class FromCloudFormation {
       return Token.asNumber(value);
     }
 
-    // in all other cases, just return the input,
-    // and let a validator handle it if it's not a number
+    // return a number, if the input can be parsed as one
+    let parsedValue;
+    if (typeof value === 'string') {
+      parsedValue = parseFloat(value);
+      if (!isNaN(parsedValue)) {
+        return parsedValue;
+      }
+    }
+
+    // otherwise, return the input
```

Comment:
...and let the validator handle the fact that it's not the type we want.

That's a pretty important detail to understanding the behavior of the parser here. At first glance this code would be wrong, but it becomes correct in combination with some other code somewhere else.